src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,219 +287,398 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
+    /// request was handled (typically by returning an error message).  If <see langword="false"/>
+    /// is returned, the request is processed normally.
+    /// <br/><br/>
+    /// By default this does not check authorization rules becuase authentication may take place within
+    /// the WebSocket connection during the ConnectionInit message.  Authorization checks for
+    /// WebSocket connections occur then, after authorization has taken place.
+    /// </summary>
+    protected virtual ValueTask<bool> HandleAuthorizeWebSocketConnectionAsync(HttpContext context, RequestDelegate next)
+        => new(false);
+
+    /// <summary>
+    /// Handles a single GraphQL request.
+    /// </summary>
     protected virtual async Task HandleRequestAsync(
         HttpContext context,
         RequestDelegate next,
-        IDictionary<string, object> userContext,
-        IList<GraphQLRequest> bodyGQLBatchRequest,
-        GraphQLRequest gqlRequest,
-        IDocumentExecuter<TSchema> executer,
-        CancellationToken cancellationToken)
+        GraphQLRequest gqlRequest)
     {
         // Normal execution with single graphql request
-        if (bodyGQLBatchRequest == null)
-        {
-            var stopwatch = ValueStopwatch.StartNew();
-            await RequestExecutingAsync(gqlRequest);
-            var result = await ExecuteRequestAsync(gqlRequest, userContext, executer, context.RequestServices, cancellationToken);
-
-            await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequest, result, stopwatch.Elapsed));
+        var userContext = await BuildUserContextAsync(context);
+        var result = await ExecuteRequestAsync(context, gqlRequest, context.RequestServices, userContext);
+        var statusCode = Options.ValidationErrorsReturnBadRequest && !result.Executed
+            ? HttpStatusCode.BadRequest
+            : HttpStatusCode.OK;
+        await WriteJsonResponseAsync(context, statusCode, result);
+    }
 
-            await WriteResponseAsync(context.Response, _serializer, cancellationToken, result);
+    /// <summary>
+    /// Handles a batched GraphQL request.
+    /// </summary>
+    protected virtual async Task HandleBatchRequestAsync(
+        HttpContext context,
+        RequestDelegate next,
+        IList<GraphQLRequest?> gqlRequests)
+    {
+        var userContext = await BuildUserContextAsync(context);
+        var results = new ExecutionResult[gqlRequests.Count];
+        if (gqlRequests.Count == 1)
+        {
+            results[0] = await ExecuteRequestAsync(context, gqlRequests[0], context.RequestServices, userContext);
         }
-        // Execute multiple graphql requests in one batch
         else
         {
-            var executionResults = new ExecutionResult[bodyGQLBatchRequest.Count];
-            for (int i = 0; i < bodyGQLBatchRequest.Count; ++i)
+            // Batched execution with multiple graphql requests
+            if (!Options.ExecuteBatchedRequestsInParallel)
```

Comment:
I would invert condition.

Reply:
Just to make extra work?!?  If you insist...

Reply:
Conditions without `!` are always easier to read than conditions with `!`. This is an additional logical operation in the head.